### PR TITLE
fix(bigquery/storage): handle retryable RESOURCE_EXHAUSTED errors in bigquery storage sample

### PR DESCRIPTION
**Pull request by esert-g**
* Adds special handling for retryable RESOURCE_EXHAUSTED errors returned from bigquery storage api, so they aren't treated as regular errors.
* Fixes retries being reset every time an error is observed, so at most 3 back-to-back errors are retried.

### CODE REVIEW

File: bigquery/bigquery_storage_quickstart/main.go [hunk 1]
```diff
@@ -224,10 +227,31 @@ func processStream(ctx context.Context, client *bqStorage.BigQueryReadClient, st
 			if err == io.EOF {
 				return nil
 			}
-			if err != nil {
-				retries++
-				if retries >= retryLimit {
-					return fmt.Errorf("processStream retries exhausted: %v", err)
+			if err == nil {
```
Comment (shollyman) [inline bigquery/bigquery_storage_quickstart/main.go:230]: Can you reorder this so the `err==nil` case is last?  As it stands, the logic checks for io.EOF, no error, then generic error (where we inspect for the exhaustion/retryinfo).

File: bigquery/bigquery_storage_quickstart/main.go [hunk 2]
```diff
@@ -224,10 +227,31 @@ func processStream(ctx context.Context, client *bqStorage.BigQueryReadClient, st
...
+				// Reset retries after a successful response.
+				retries = 0
+			} else {
+				// If there is an error, check whether it is a retryable error with a retry delay and sleep instead of increasing retries count.
+				var retryDelayDuration time.Duration
+				if errorStatus, ok := status.FromError(err); ok && errorStatus.Code() == codes.ResourceExhausted {
```
Comment (shollyman) [inline bigquery/bigquery_storage_quickstart/main.go:236]: Do we ever anticipate the backend adding RetryInfo to responses other than ResourceExhausted?  If so, perhaps we don't special case the `ResourceExhausted`.

**Review by shollyman — COMMENTED** [review]
Thanks for the PR, a couple minor items then we should be able to get this merged.

File: bigquery/bigquery_storage_quickstart/main.go (hunk 1 above)
Comment (esert-g) [reply bigquery/bigquery_storage_quickstart/main.go]: Done.

File: bigquery/bigquery_storage_quickstart/main.go (hunk 2 above)
Comment (esert-g) [reply bigquery/bigquery_storage_quickstart/main.go:232]: BigQuery storage backend will only return a RetryInfo with ResourceExhausted.

**Review by shollyman — APPROVED** [review]

File: bigquery/bigquery_storage_quickstart/main.go [hunk 3]
```diff
@@ -223,15 +226,35 @@ func processStream(ctx context.Context, client *bqStorage.BigQueryReadClient, st
 			r, err := rowStream.Recv()
 			if err == io.EOF {
 				return nil
-			}
-			if err != nil {
-				retries++
-				if retries >= retryLimit {
-					return fmt.Errorf("processStream retries exhausted: %v", err)
+			} else if err != nil {
```
Comment (enocom) [inline bigquery/bigquery_storage_quickstart/main.go:229]: The else here is superfluous with the return above.

File: bigquery/bigquery_storage_quickstart/main.go [hunk 4]
```diff
@@ -223,15 +226,35 @@ func processStream(ctx context.Context, client *bqStorage.BigQueryReadClient, st
...
+				// If there is an error, check whether it is a retryable error with a retry delay and sleep instead of increasing retries count.
```
Comment (enocom) [inline bigquery/bigquery_storage_quickstart/main.go:230]: Please wrap this comment to under 100 characters.

**Review by enocom — CHANGES_REQUESTED** [review]
A couple minor things

File: bigquery/bigquery_storage_quickstart/main.go (hunk 3 above)
Comment (esert-g) [reply bigquery/bigquery_storage_quickstart/main.go]: Done

File: bigquery/bigquery_storage_quickstart/main.go (hunk 4 above)
Comment (esert-g) [reply bigquery/bigquery_storage_quickstart/main.go]: Done

**Review by enocom — APPROVED** [review]

**Pull request merged**